src/components/implementation/no_interface/new_boot/ppos_booter.c

```diff
@@ -0,0 +1,344 @@
+	struct cos_component_information *ci;
+
+	start_addr = (char *)(new_comp_cap_info[spdid].vaddr_mapped_in_booter);
+	init_daddr = cobj_sect_get(h, 0)->vaddr;
+	for (i = 0 ; i < h->nsect ; i++) {
+		struct cobj_sect *sect;
+		vaddr_t dest_daddr;
+		char *lsrc, *dsrc;
+		int left, dest_doff;
+
+		sect       = cobj_sect_get(h, i);
+		/* virtual address in the destination address space */
+		dest_daddr = sect->vaddr;
+		/* where we're copying from in the cobj */
+		lsrc       = cobj_sect_contents(h, i);
+		/* how much is left to copy? */
+		left       = cobj_sect_size(h, i);
+
+		/* Initialize memory. */
+		if (!(sect->flags & COBJ_SECT_KMEM) &&
+		    (first_time || !(sect->flags & COBJ_SECT_INITONCE))) {
+			if (sect->flags & COBJ_SECT_ZEROS) {
+      				memset(start_addr + (dest_daddr - init_daddr), 0, left);
+			} else {
+				memcpy(start_addr + (dest_daddr - init_daddr), lsrc, left);
+			}
+		}
+
+		if (sect->flags & COBJ_SECT_CINFO) {
+			assert(left == PAGE_SIZE);
+			assert(comp_info == dest_daddr);
+			boot_process_cinfo(h, spdid, boot_spd_end(h), start_addr + (comp_info-init_daddr), comp_info);
+			ci = (struct cos_component_information*)(start_addr + (comp_info-init_daddr));
+			new_comp_cap_info[h->id].upcall_entry = ci->cos_upcall_entry;
+			
+			vaddr_t begin =start_addr + ((ci->cos_upcall_entry) - init_daddr); 
+		}
+	}
+
+	return 0;
+}
+
+int
+boot_comp_map(struct cobj_header *h, spdid_t spdid, vaddr_t comp_info, pgtblcap_t pt)
+{
+	boot_comp_map_memory(h, spdid, pt);
+	boot_comp_map_populate(h, spdid, comp_info, 1);
+	return 0;
+}
+			
+static void
+boot_create_cap_system(void)
+{
+	unsigned int i;
+
+	for (i = 0 ; hs[i] != NULL ; i++) {
+		struct cobj_header *h;
+		struct cobj_sect *sect;
+		captblcap_t ct;
+		pgtblcap_t pt;
+		spdid_t spdid;
+		vaddr_t ci = 0;
+		
+		h = hs[i];
+		spdid = h->id;
+		printc("Initializing Comp: %s\n", h->name);	
+		
+		sect = cobj_sect_get(h, 0);
+		new_comp_cap_info[spdid].addr_start = sect->vaddr;
+
+		boot_compinfo_init(spdid, &ct, &pt, sect->vaddr);	
+
+		if (boot_spd_symbs(h, spdid, &ci)) BUG();
+		if (boot_comp_map(h, spdid, ci, pt))   BUG();
+
+		boot_newcomp_create(spdid, ct, pt);
+		
+		printc("Comp %d (%s) created @ %x!\n", h->id, h->name, sect->vaddr);
+	}
+
+	return;
+}
+
+static void
+boot_init_sched(void)
+{
+	int i;
+	
+	for (i = 0 ; i < MAX_NUM_SPDS ; i++) schedule[i] = NULL;
+	sched_cur = 0;
+}
+
+			
+void
+boot_init_ndeps(void)
+{
+	int i;
+
+	for (i = 0 ; deps[i].server ; i++) ;
+	ndeps = i;
+}
+
+void 
+cos_init(void)
+{
+ 	struct cobj_header *h;
+	int num_cobj, i;
+
+	printc("Booter for new kernel\n");
+
+	h         = (struct cobj_header *)cos_comp_info.cos_poly[0];
+	num_cobj  = (int)cos_comp_info.cos_poly[1];
+
+	deps      = (struct deps *)cos_comp_info.cos_poly[2];
+	boot_init_ndeps();
+
+	init_args = (struct component_init_str *)cos_comp_info.cos_poly[3];
+	init_args++;
+
+	boot_sched = (unsigned int *)cos_comp_info.cos_poly[4];
+	
+	boot_find_cobjs(h, num_cobj);
+
```

Comment:
Can remove the empty lines between these functions.  They aren't really "separate paragraphs".